examples/ecs/index.rs

```diff
@@ -0,0 +1,76 @@
+//! Demonstrates how to query for a component _value_ using indexes.
+
+use bevy::prelude::*;
+
+// To query by component value, first we need to ensure our component is suitable
+// for indexing.
+//
+// The hard requirements are:
+// * Immutability
+// * Eq + Hash + Clone
+
+#[derive(Component, PartialEq, Eq, Hash, Clone, Debug)]
+#[component(immutable)]
+struct TilePosition {
+    // It's strongly recommended to keep the range of possible values for an index component small.
+    // Each unique value in use at the same time will require a ComponentId, so large spaces like `f32` are highly discouraged.
+    x: i8,
+    y: i8,
+}
+
+fn main() {
+    App::new()
+        // Simply call `add_index` to start indexing a component.
+        // Make sure to call this _before_ any components are spawned, otherwise
+        // the index will miss those entities!
+        .add_index::<TilePosition>()
```

Comment:
Maybe it would be worthwhile to change `Index<C>` from a `Resource` to a `Component` `Index<C>` that can be added on any entity that we want to be indexable by C.

This lets us:
- ignore some entities that we don't want to get indexed (cheaper indexation). In your version they would still get indexed and the user would have to filter them out in their `QueryByIndex`
- have an easy way to remove an entity from an index (by removing the `Index<C>` component)

For example I might only want to index the `Transform` of only a few entities (player entities), but not of sprites, walls, etc.

Maybe both version can co-exist, because your version has ergonomic benefits (users don't have to add Index<C> on every entity that needs to be indexed)